tiledb/sm/query/readers/aggregators/mean_aggregator.cc

```diff
@@ -0,0 +1,269 @@
+ public:
+  explicit MeanAggregatorStatusException(const std::string& message)
+      : StatusException("MeanAggregator", message) {
+  }
+};
+
+template <typename T>
+MeanAggregator<T>::MeanAggregator(const FieldInfo field_info)
+    : field_info_(field_info)
+    , sum_(0)
+    , count_(0)
+    , validity_value_(
+          field_info_.is_nullable_ ? std::make_optional(0) : nullopt)
+    , sum_overflowed_(false) {
+  if (field_info_.var_sized_) {
+    throw MeanAggregatorStatusException(
+        "Mean aggregates must not be requested for var sized attributes.");
+  }
+
+  if (field_info_.cell_val_num_ != 1) {
+    throw MeanAggregatorStatusException(
+        "Mean aggregates must not be requested for attributes with more than "
+        "one value.");
+  }
+}
+
+template <typename T>
+void MeanAggregator<T>::validate_output_buffer(
+    std::string output_field_name,
+    std::unordered_map<std::string, QueryBuffer>& buffers) {
+  if (buffers.count(output_field_name) == 0) {
+    throw MeanAggregatorStatusException("Result buffer doesn't exist.");
+  }
+
+  auto& result_buffer = buffers[output_field_name];
+  if (result_buffer.buffer_ == nullptr) {
+    throw MeanAggregatorStatusException(
+        "Mean aggregates must have a fixed size buffer.");
+  }
+
+  if (result_buffer.buffer_var_ != nullptr) {
+    throw MeanAggregatorStatusException(
+        "Mean aggregates must not have a var buffer.");
+  }
+
+  if (result_buffer.original_buffer_size_ != 8) {
+    throw MeanAggregatorStatusException(
+        "Mean aggregates fixed size buffer should be for one element only.");
+  }
+
+  bool exists_validity = result_buffer.validity_vector_.buffer();
+  if (field_info_.is_nullable_) {
+    if (!exists_validity) {
+      throw MeanAggregatorStatusException(
+          "Mean aggregates for nullable attributes must have a validity "
+          "buffer.");
+    }
+
+    if (*result_buffer.validity_vector_.buffer_size() != 1) {
+      throw MeanAggregatorStatusException(
+          "Mean aggregates validity vector should be for one element only.");
+    }
+  } else {
+    if (exists_validity) {
+      throw MeanAggregatorStatusException(
+          "Mean aggregates for non nullable attributes must not have a "
+          "validity "
+          "buffer.");
+    }
+  }
+}
+
+template <typename T>
+void MeanAggregator<T>::aggregate_data(AggregateBuffer& input_data) {
+  tuple<double, uint64_t, optional<uint8_t>> res{0, 0, nullopt};
+
+  bool overflow = false;
+  try {
+    if (input_data.is_count_bitmap()) {
+      res = mean<uint64_t>(input_data);
+    } else {
+      res = mean<uint8_t>(input_data);
+    }
+  } catch (std::overflow_error&) {
+    overflow = true;
+  }
+
+  {
+    // This might be called on multiple threads, the final result stored in sum_
+    // should be computed in a thread safe manner. The mutex also protects
+    // sum_overflowed_ which indicates when the sum has overflowed.
+    std::unique_lock lock(mean_mtx_);
```

Comment:
Why do we have a mutex here? This seems to be mixing concerns: there should be no parallelism at the individual compute kernel level. Parallelism should happen one step above, composed over chunks if necessary. Not passed through a single class with a guard like this.

x-ref https://github.com/TileDB-Inc/TileDB/pull/4113#discussion_r1269402199

Reply:
We have parallel fors in the readers that call the aggregate data function for various spans of data on multiple threads... The aggregator holds the final value of the aggregation, here being the sum... I'm not sure how we can protect the single data collection point by anything other than a mutex?

I though https://github.com/TileDB-Inc/TileDB/pull/4113#discussion_r1269402199 was asking to precise what the mutex was protecting and not the actual design.